jetty-core/jetty-xml/src/main/java/org/eclipse/jetty/xml/EnvironmentBuilder.java

```diff
@@ -13,54 +13,187 @@
 
 package org.eclipse.jetty.xml;
 
-import java.io.IOException;
+import java.lang.module.Configuration;
+import java.lang.module.ModuleFinder;
+import java.net.MalformedURLException;
 import java.net.URL;
 import java.net.URLClassLoader;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.ArrayList;
+import java.util.LinkedHashSet;
 import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
 
 import org.eclipse.jetty.util.annotation.Name;
 import org.eclipse.jetty.util.component.Environment;
-import org.eclipse.jetty.util.resource.Resource;
-import org.eclipse.jetty.util.resource.ResourceFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * A Builder of {@link Environment}s intended to be used in XML
- * files generated by <code>start.jar</code>.
- *
+ * files processed by {@code start.jar}.
  */
 public class EnvironmentBuilder
 {
+    private static final Logger LOG = LoggerFactory.getLogger(EnvironmentBuilder.class);
+
     private final String _name;
-    private final List<URL> _classpath = new ArrayList<>();
+    private final List<Path> _classPath = new ArrayList<>();
+    private final List<Path> _modulePath = new ArrayList<>();
+    private final List<String> _jpmsAdds = new ArrayList<>();
+    private final List<String> _jpmsOpens = new ArrayList<>();
+    private final List<String> _jpmsExports = new ArrayList<>();
+    private final List<String> _jpmsReads = new ArrayList<>();
 
     public EnvironmentBuilder(@Name("name") String name)
     {
         _name = name;
     }
 
-    public void addModulePath(String arg)
+    public void addClassPath(String... classPaths)
     {
-        throw new UnsupportedOperationException();
+        for (String classPath : classPaths)
+        {
+            _classPath.add(Paths.get(classPath));
+        }
     }
 
-    public void addClassPath(String... classPaths)
+    public void addModulePath(String modulePath)
     {
-        for (String classPath : classPaths)
+        _modulePath.add(Paths.get(modulePath));
+    }
+
+    void addModules(String modules)
+    {
+        _jpmsAdds.addAll(List.of(modules.split(",")));
+    }
+
+    void patchModule(String patch)
+    {
+        // Not supported by the ModuleLayer.Controller APIs.
+    }
+
+    void addOpens(String opens)
+    {
+        _jpmsOpens.add(opens);
+    }
+
+    void addExports(String exports)
+    {
+        _jpmsExports.add(exports);
+    }
+
+    void addReads(String reads)
+    {
+        _jpmsReads.add(reads);
+    }
+
+    public Environment build() throws Exception
```

Comment:
That now looks complicated enough to be worthy of a test. But can we use junit to write JPMS tests?